### PR TITLE
fix timeout to allow fake timers

### PR DESCRIPTION
**Pull request by mjhm**
@charlierudolph This is the fix to allow `sinon.useFakeTimers` or similar time mocking to work. https://github.com/cucumber/cucumber-js/issues/535

I'm not sure exactly where to plug in new tests for it though.

I have my own example test repo at https://github.com/mjhm/cucumber-test which demonstrates the problem using 0.10.2 vs this update.

### CODE REVIEW

**Comment by charlierudolph** [comment]
Maybe throw in a feature test that uses sinon.useFakeTimers or stubs out one of the methods?

**Comment by mjhm** [comment]
@charlierudolph Added feature tests.

File: features/fake_time.feature [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        });
+      };
+
+      module.exports = stepDefinitions
+      """
+      When I run cucumber.js with `--strict`
+      Then it passes
+
+
+  Scenario: real time is restored
```
Comment (charlierudolph) [inline features/fake_time.feature]: I'm not certain this is test is really needed. The name makes me think that real time was mocked but it never was.

File: features/fake_time.feature [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+      module.exports = stepDefinitions
+      """
+
+      When I run cucumber.js with `--strict`
+      Then it passes
+
+
+  Scenario: faked time doesn't trigger the test runner timeout
+
```
Comment (charlierudolph) [inline features/fake_time.feature]: unnecessary whitespace

File: features/fake_time.feature [hunk 3]
```diff
@@ -0,0 +1,129 @@
+Feature: Fake Time
```
Comment (charlierudolph) [inline features/fake_time.feature]: Can you please add a description for this feature.

File: lib/cucumber/util/real_time.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (charlierudolph) [inline lib/cucumber/util/real_time.js]: unnecessary whitespace.

File: lib/cucumber/support_code/step_definition.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
+var realTime = require('../util/real_time');
```
Comment (charlierudolph) [inline lib/cucumber/support_code/step_definition.js]: Instead of requiring this directly, please set `Cucumber.Util.RealTime` in `lib/cucumber/util.js`

**Comment by charlierudolph** [comment]
Thanks @mjhm. Feature tests look great. Just a couple things to cleanup and this is good to go.

File: features/fake_time.feature (hunk 1 above)
Comment (mjhm) [inline features/fake_time.feature]: The point of this test is to show that after running the first scenario in a faked time context (note the `@fakedTime` tag), then the second scenario runs with real time restored (no `@fakedTime` tag).

**Comment by mjhm** [comment]
Here you go @charlierudolph

**Comment by mjhm** [comment]
Ping @charlierudolph

File: features/fake_time.feature (hunk 1 above)
Comment (charlierudolph) [inline features/fake_time.feature]: Yeah, but to cucumber-js real time was never faked. To me this is just testing that you correctly ran this.clock.restore() in your after hook. Not running that doesn't effect whether or not cucumber-js works.

File: features/fake_time.feature [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+      this.Before({tags: ['@fakedTime']}, function(scenario, callback) {
+        this.clock = sinon.useFakeTimers();
+        callback();
+      });
+
+      this.After({tags: ['@fakedTime']}, function(scenario, callback) {
+        this.clock.restore();
+        callback();
+      });
```
Comment (charlierudolph) [inline features/fake_time.feature]: you can just omit the callback in each of these since they are synchronous. Thus you can omit scenario as well

File: features/fake_time.feature [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      var sinon = require('sinon')
+
+      stepDefinitions = function() {
+        this.Given(/^a faked time step$/, function (done) {
+          var testFunction = sinon.stub();
+          setTimeout(testFunction, 100);
+          assert(!testFunction.called);
+          this.clock.tick(101);
+          assert(testFunction.called);
+          done()
```
Comment (charlierudolph) [inline features/fake_time.feature]: you can omit the callback as this is synchronous.
